src/esphome/sensor/sds011_component.cpp

```diff
@@ -0,0 +1,243 @@
+//   - https://cdn.sparkfun.com/assets/parts/1/2/2/7/5/Laser_Dust_Sensor_Control_Protocol_V1.3.pdf
+
+#include "esphome/defines.h"
+
+#ifdef USE_SDS011
+
+#include "esphome/log.h"
+#include "esphome/sensor/sds011_component.h"
+
+ESPHOME_NAMESPACE_BEGIN
+
+namespace sensor {
+
+static const char *TAG = "sensor.sds011";
+static const uint8_t SDS011_MSG_REQUEST_LENGTH = 19;
+static const uint8_t SDS011_MSG_RESPONSE_LENGTH = 10;
+static const uint8_t SDS011_DATA_REQUEST_LENGTH = 15;
+static const uint8_t SDS011_DATA_RESPONSE_LENGTH = 6;
+static const uint8_t SDS011_MSG_HEAD = 0xaa;
+static const uint8_t SDS011_MSG_TAIL = 0xab;
+static const uint8_t SDS011_COMMAND_ID_REQUEST = 0xb4;
+static const uint8_t SDS011_COMMAND_ID_RESPONSE = 0xc5;
+static const uint8_t SDS011_COMMAND_ID_DATA = 0xc0;
+static const uint8_t SDS011_COMMAND_REPORT_MODE = 0x02;
+static const uint8_t SDS011_COMMAND_QUERY_DATA = 0x04;
+static const uint8_t SDS011_COMMAND_SET_DEVICE_ID = 0x05;
+static const uint8_t SDS011_COMMAND_SLEEP = 0x06;
+static const uint8_t SDS011_COMMAND_FIRMWARE = 0x07;
+static const uint8_t SDS011_COMMAND_PERIOD = 0x08;
+static const uint8_t SDS011_GET_MODE = 0x00;
+static const uint8_t SDS011_SET_MODE = 0x01;
+static const uint8_t SDS011_MODE_REPORT_ACTIVE = 0x00;
+static const uint8_t SDS011_MODE_REPORT_QUERY = 0x01;
+static const uint8_t SDS011_MODE_SLEEP = 0x00;
+static const uint8_t SDS011_MODE_WORK = 0x01;
+
+SDS011Component::SDS011Component(UARTComponent *parent, uint32_t update_interval, bool query_mode, bool rx_mode_only)
+    : UARTDevice(parent),
+      update_interval_(update_interval),
+      query_mode_(query_mode),
+      rx_mode_only_(rx_mode_only),
+      force_query_(false) {}
+
+void SDS011Component::setup() {
+  if (this->get_rx_mode_only()) {
+    if (this->get_query_mode()) {
+      ESP_LOGE(TAG, "SDS011 component can not work in rx-only mode and query mode.");
+    }
+    return;
+  }
+
+  uint8_t command_data[SDS011_DATA_REQUEST_LENGTH] = {0};
+  command_data[0] = SDS011_COMMAND_REPORT_MODE;
+  command_data[1] = SDS011_SET_MODE;
+  command_data[2] = this->query_mode_ ? SDS011_MODE_REPORT_QUERY : SDS011_MODE_REPORT_ACTIVE;
+  command_data[13] = 0xff;
+  command_data[14] = 0xff;
+  this->sds011_write_command_(command_data);
+
+  uint8_t update_interval_minutes = this->get_update_interval() / 1000.0f / 60.0f;
+  this->set_update_interval((uint32_t) update_interval_minutes * 1000 * 60);
+  if (update_interval_minutes > 30) {
+    ESP_LOGW(TAG, "Update interval is longer than 30 min.");
+  }
+
+  command_data[0] = SDS011_COMMAND_PERIOD;
+  command_data[1] = SDS011_SET_MODE;
+  command_data[2] = update_interval_minutes;
+  command_data[13] = 0xff;
+  command_data[14] = 0xff;
+  this->sds011_write_command_(command_data);
+
+  if (this->get_query_mode()) {
+    this->update();
+    this->set_timeout("delayed-update", 35000, [this]() {
+      this->set_interval("update", this->get_update_interval(), [this]() { this->update(); });
```

Comment:
We should not override the PollingComponent's update `set_interval`.

What is the reason for this delayed second update?

Reply:
> We should not override the PollingComponent's update set_interval.

What do you mean with override? This component does not inherit from the `PollingComponent` class. However, with `query_mode`, it mimics the `PollingComponent` class. If you don't like it, I will change it.

> What is the reason for this delayed second update?

The query command only ask for the last measurement value and do not force a measurement. For working periods (update_interval >= 1min), the sensor need 30s for the measurement (mainly to get fresh air from the fan). So for a power cycle of esp and sensor, we have to wait 30s to get a valid value. In the case, the esp rebooted or was in deepsleep (esp8266), we can read a valid value with the first `update()`.

Reply:
> What do you mean with override? This component does not inherit from the PollingComponent class. However, with query_mode, it mimics the PollingComponent class. If you don't like it, I will change it.

Oh, I assumed it was overriding `PollingComponent`. In that case we should override `PollingComponent` and `return` in update if we're not in query mode.

> The query command only ask for the last measurement value and do not force a measurement. For working periods (update_interval >= 1min), the sensor need 30s for the measurement (mainly to get fresh air from the fan). So for a power cycle of esp and sensor, we have to wait 30s to get a valid value. In the case, the esp rebooted or was in deepsleep (esp8266), we can read a valid value with the first update().

If the user only sees this on the first time they plug the device in I think it's not a problem. Deep sleep is likely also not that important here because the sensor surely uses tons of power that would quickly drain all batteries.

Reply:
As far as I can see, when we switch to `PollingComponent`, we can not delay the first query and always get the value of the previous measurement (which is several minutes old). IMHO, `query_mode` only makes sense for deep_sleep

Reply:
That is true, however:

- `deep_sleep` does not really make sense with this integration I think. I'm pretty sure the sensor uses tons of energy even when in idle mode (might be wrong about this). And when it's in non-idle mode it surely uses a ton of energy too - so you'd only get at most a week of deep sleep uptime
- While delaying the initial reading might be the correct way of doing it, I think it's not that important, because: It only ever happens when the user power-cycles the setup or on initial data - IMO those two cases happen so seldomly that it's ok if we do not handle those correctly